src/OpenTelemetry.Exporter.Prometheus.HttpListener/PrometheusHttpListener.cs

```diff
@@ -105,13 +105,23 @@ public void Stop()
             }
         }
 
-        /// <inheritdoc/>
         public void Dispose()
         {
-            if (this.httpListener != null && this.httpListener.IsListening)
+            this.Dispose(true);
+        }
+
+        private void Dispose(bool disposing)
+        {
+            if (!this.disposed)
             {
-                this.Stop();
-                this.httpListener.Close();
+                if (disposing && this.httpListener != null && this.httpListener.IsListening)
+                {
+                    this.Stop();
+                    this.httpListener.Close();
+                    this.Dispose();
```

Comment:
* Not totally sure if `this.Stop()` should be part of `disposing` check or not. If you make a mistake and don't call Dispose and never Stop the listener does the process hang?
* Won't `this.Dispose()` call here cause an infinite loop?

Reply:
I've moved out `this.Stop()` out of the if check and also removed this.Dipose() which is buggy.

Regarding this:
> If you make a mistake and don't call Dispose and never Stop the listener does the process hang?

-> I think the destructor will take care of it?

Reply:
We would have to do something like this...

```csharp
        ~PrometheusHttpListener()
        {
            this.Stop();
        }
```

But that is kind of frowned upon.

I played with it a bit. Today when MeterProvider is disposed, exporter is disposed, http listener is disposed here:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/be588bdcd3a3006a61f9f8f4c957ef4ba66f0a8a/src/OpenTelemetry.Exporter.Prometheus/PrometheusExporter.cs#L85

That is missing in the new flow. Actually, in the new flow there is no way to stop/dispose the listener at all because it never escapes the scope of `AddPrometheusHttpListener`.

I just pushed a change to the branch that ties the disposal of `PrometheusHttpListener` to `PrometheusExporter` which basically restores the old behavior. Let me know what you think about that.

Reply:
I'm not sure whether [the binding](https://github.com/open-telemetry/opentelemetry-dotnet/blob/17fa0b73bd4dd4950236255aad20dd8394ffa2e7/src/OpenTelemetry.Exporter.Prometheus.HttpListener/PrometheusExporterHttpListenerMeterProviderBuilderExtensions.cs#:~:text=exporter.OnDispose%20%3D%20()%20%3D%3E%20listener.Dispose()%3B) should be called out explicitly as I learned that:
> If disposable object x contains or “wraps” or “possesses” disposable object y, x’s Dispose method automatically calls y’s Dispose method—unless instructed otherwise.
![image](https://user-images.githubusercontent.com/16979322/178384452-401611dc-363e-40f9-8f49-984135d38faf.png)
From: [C# 5.0 in a Nutshell, 5th Edition](https://learning.oreilly.com/library/view/c-5-0-in/9781449334192/)
[Chapter 12. Disposal and Garbage Collection](https://learning.oreilly.com/library/view/c-5-0-in/9781449334192/ch12.html#:-:text=According%20to%20rule%203,otherwise%20in%0A%20%20%20%20%20%20the%20constructor.)

Other than that, 17fa0b73bd4dd4950236255aad20dd8394ffa2e7 looks good to me!
Thank you Blanch! :smiley:

Reply:
Chatted with Utkarsh offline and figured out that only .NET framework classes will do this auto-binding thing. We're all good now :-).